File: NowCoder/Contest/牛客周赛/牛客周赛 Round 75/小红的双生英雄.py
N, C, M = map(int, input().split())
heroes = [tuple(map(int, input().split())) for _ in range(N)]
pairs = [tuple(map(int, input().split())) for _ in range(M)]

groups = []
is_pair = [False] * N
for u, v, w in pairs:
    u -= 1
    v -= 1
    is_pair[u] = True
    is_pair[v] = True
    groups.append((u, v, w))
for i in range(N):
    if not is_pair[i]:
        groups.append((i, i, 0))

n = len(groups)
f = [[-float('inf')] * (5) for _ in range(C + 1)]
f[0][0] = 0
for u, v, w in groups:
    cost_u, w_u = heroes[u]
    cost_v, w_v = heroes[v]
    cost_both = cost_u + cost_v
    w_both = w_u + w_v + w

    new_f = [row[:] for row in f]
    for j in range(C + 1):
        for k in range(5):
            if f[j][k] == -float('inf'):
                continue
            new_f[j][k] = max(new_f[j][k], f[j][k])
            if j + cost_u <= C and k + 1 <= 4:
                new_f[j + cost_u][k + 1] = max(new_f[j + cost_u][k + 1], f[j][k] + w_u)
            if u != v and j + cost_v <= C and k + 1 <= 4:
                new_f[j + cost_v][k + 1] = max(new_f[j + cost_v][k + 1], f[j][k] + w_v)
            if u != v and j + cost_both <= C and k + 2 <= 4:
                new_f[j + cost_both][k + 2] = max(new_f[j + cost_both][k + 2], f[j][k] + w_both)
    f = new_f

ans = 0
for j in range(C + 1):
    for k in range(5):
        ans = max(ans, f[j][k])
print(ans)

File: NowCoder/Contest/牛客周赛/牛客周赛 Round 75/小红的双生英雄.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const LL INF = 1e18;
#define endl '\n'

struct Hero {
    int cost, w;
};

struct Group {
    int u, v, w;
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int N, C, M, cost, u, v, w;
    cin >> N >> C >> M;
    vector<Hero> heroes(N);
    for (int i = 0; i < N; i++) {
        cin >> cost >> w;
        heroes[i] = {cost, w};
    }
    vector<Group> groups;
    vector<bool> is_pair(N, false);
    for (int i = 0; i < M; i++) {
        cin >> u >> v >> w;
        u--; v--;
        is_pair[u] = true;
        is_pair[v] = true;
        groups.push_back({u, v, w});
    }
    for (int i = 0; i < N; i++) {
        if (!is_pair[i]) {
            groups.push_back({i, i, 0});
        }
    }
    int n = groups.size();
    vector<vector<vector<LL>>> memo(n + 1, vector<vector<LL>>(C + 1, vector<LL>(5, -INF)));
    auto dfs = [&](auto &&dfs, int i, int j, int k) -> LL {
        if (j > C || k > 4) return -INF;
        if (i == n || j == C || k == 4) return 0;
        LL &res = memo[i][j][k];
        if (res != -INF) return res;
        res = dfs(dfs, i + 1, j, k);
        auto [u, v, w] = groups[i];
        if (u != v) {
            auto [cost_u, w_u] = heroes[u];
            auto [cost_v, w_v] = heroes[v];
            res = max({res, 
                      dfs(dfs, i + 1, j + cost_u, k + 1) + w_u,
                      dfs(dfs, i + 1, j + cost_v, k + 1) + w_v,
                      dfs(dfs, i + 1, j + cost_u + cost_v, k + 2) + w_u + w_v + w});
        } else {
            auto [cost, w] = heroes[u];
            res = max(res, dfs(dfs, i + 1, j + cost, k + 1) + w);
        }
        return res;
    };
    cout << dfs(dfs, 0, 0, 0) << endl;
    return 0;
}